File: .agents/skills/yolo/scripts/polyglot-detector.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Universal Polyglot Language & File Type Detector & Parser for YOLO+
 * Inspired by GitHub Linguist, Google Magika, go-enry, and Tree-sitter heuristics.
 *
 * Capabilities:
 * - 60+ programming languages & file types detected natively.
 * - Multi-stage resolution:
 *   1. Filename / Exact match (e.g. Dockerfile, CMakeLists.txt, Makefile, etc.)
 *   2. Extension mapping (primary & aliases)
 *   3. Shebang / Content modeline inspection (e.g. #!/usr/bin/env python, <?php, etc.)
 *   4. Content heuristic classification (Bayesian/pattern inspired for ambiguous extensions like .m, .h, .pl)
 * - AST / Structural parsing extraction:
 *   Extracts functions, classes, interfaces, imports/modules, symbols, line counts, comment ratios.
 * - Universal build & test gate resolution for every detected language.
 */

export interface LanguageDefinition {
  id: string;
  name: string;
  category: "programming" | "data" | "markup" | "prose" | "config" | "binary";
  extensions: string[];
  filenames?: string[];
  shebangs?: string[];
  testRunner?: string;
  buildGate?: string;
  mimeType: string;
}

export interface ParseResult {
  language: string;
  category: string;
  mimeType: string;
  isBinary: boolean;
  totalLines: number;
  codeLines: number;
  commentLines: number;
  blankLines: number;
  imports: string[];
  declarations: {
    classes: string[];
    functions: string[];
    interfaces: string[];
    constants: string[];
  };
  validationGate?: string;
}

export const LANGUAGE_REGISTRY: LanguageDefinition[] = [
  {
    id: "rust",
    name: "Rust",
    category: "programming",
    extensions: [".rs"],
    filenames: ["Cargo.toml"],
    testRunner: "cargo test",
    buildGate: "cargo check --workspace --all-targets",
    mimeType: "text/x-rust"
  },
  {
    id: "typescript",
    name: "TypeScript",
    category: "programming",
    extensions: [".ts", ".tsx", ".mts", ".cts"],
    filenames: ["tsconfig.json"],
    shebangs: ["bun", "deno", "ts-node"],
    testRunner: "bun test",
    buildGate: "tsc --noEmit",
    mimeType: "application/typescript"
  },
  {
    id: "javascript",
    name: "JavaScript",
    category: "programming",
    extensions: [".js", ".jsx", ".mjs", ".cjs"],
    filenames: ["package.json"],
    shebangs: ["node", "bun"],
    testRunner: "bun test || npm test",
    buildGate: "node --check",
    mimeType: "application/javascript"
  },
  {
    id: "python",
    name: "Python",
    category: "programming",
    extensions: [".py", ".pyw", ".pyi"],
    filenames: ["pyproject.toml", "requirements.txt", "Pipfile"],
    shebangs: ["python", "python3", "uv run"],
    testRunner: "pytest || python -m unittest",
    buildGate: "python -m py_compile",
    mimeType: "text/x-python"
  },
  {
    id: "csharp",
    name: "C#",
    category: "programming",
    extensions: [".cs", ".csx"],
    testRunner: "dotnet test",
    buildGate: "dotnet build",
    mimeType: "text/x-csharp"
  },
  {
    id: "cpp",
    name: "C++",
    category: "programming",
    extensions: [".cpp", ".cxx", ".cc", ".hpp", ".hxx", ".hh"],
    filenames: ["CMakeLists.txt"],
    testRunner: "ctest || make test",
    buildGate: "cmake --build . || make",
    mimeType: "text/x-c++src"
  },
  {
    id: "c",
    name: "C",
    category: "programming",
    extensions: [".c", ".h"],
    filenames: ["Makefile"],
    testRunner: "make test",
    buildGate: "make",
    mimeType: "text/x-csrc"
  },
  {
    id: "assembly",
    name: "Assembly",
    category: "programming",
    extensions: [".asm", ".s", ".nasm"],
    buildGate: "nasm -f elf64 || as",
    mimeType: "text/x-asm"
  },
  {
    id: "go",
    name: "Go",
    category: "programming",
    extensions: [".go"],
    filenames: ["go.mod"],
    testRunner: "go test ./...",
    buildGate: "go build ./...",
    mimeType: "text/x-go"
  },
  {
    id: "java",
    name: "Java",
    category: "programming",
    extensions: [".java"],
    filenames: ["pom.xml", "build.gradle"],
    testRunner: "mvn test || ./gradlew test",
    buildGate: "javac",
    mimeType: "text/x-java"
  },
  {
    id: "kotlin",
    name: "Kotlin",
    category: "programming",
    extensions: [".kt", ".kts"],
    testRunner: "./gradlew test",
    buildGate: "kotlinc",
    mimeType: "text/x-kotlin"
  },
  {
    id: "swift",
    name: "Swift",
    category: "programming",
    extensions: [".swift"],
    filenames: ["Package.swift"],
    testRunner: "swift test",
    buildGate: "swift build",
    mimeType: "text/x-swift"
  },
  {
    id: "ruby",
    name: "Ruby",
    category: "programming",
    extensions: [".rb", ".rake"],
    filenames: ["Gemfile", "Rakefile"],
    shebangs: ["ruby"],
    testRunner: "bundle exec rspec",
    mimeType: "text/x-ruby"
  },
  {
    id: "php",
    name: "PHP",
    category: "programming",
    extensions: [".php", ".phtml"],
    filenames: ["composer.json"],
    shebangs: ["php"],
    testRunner: "./vendor/bin/phpunit",
    mimeType: "application/x-httpd-php"
  },
  {
    id: "shell",
    name: "Shell",
    category: "programming",
    extensions: [".sh", ".bash", ".zsh"],
    shebangs: ["bash", "sh", "zsh"],
    testRunner: "bats test",
    buildGate: "bash -n",
    mimeType: "application/x-sh"
  },
  {
    id: "powershell",
    name: "PowerShell",
    category: "programming",
    extensions: [".ps1", ".psm1", ".psd1"],
    shebangs: ["pwsh", "powershell"],
    testRunner: "Invoke-Pester",
    mimeType: "application/x-powershell"
  },
  {
    id: "html",
    name: "HTML",
    category: "markup",
    extensions: [".html", ".htm"],
    mimeType: "text/html"
  },
  {
    id: "css",
    name: "CSS",
    category: "markup",
    extensions: [".css", ".scss", ".sass", ".less"],
    mimeType: "text/css"
  },
  {
    id: "json",
    name: "JSON",
    category: "data",
    extensions: [".json", ".jsonc", ".jsonl"],
    mimeType: "application/json"
  },
  {
    id: "yaml",
    name: "YAML",
    category: "data",
    extensions: [".yaml", ".yml"],
    mimeType: "text/yaml"
  },
  {
    id: "toml",
    name: "TOML",
    category: "data",
    extensions: [".toml"],
    mimeType: "application/toml"
  },
  {
    id: "markdown",
    name: "Markdown",
    category: "prose",
    extensions: [".md", ".markdown", ".mdown"],
    mimeType: "text/markdown"
  },
  {
    id: "sql",
    name: "SQL",
    category: "data",
    extensions: [".sql"],
    mimeType: "application/sql"
  },
  {
    id: "dockerfile",
    name: "Dockerfile",
    category: "config",
    extensions: [".dockerfile"],
    filenames: ["Dockerfile", "Containerfile"],
    mimeType: "text/x-dockerfile"
  },
  {
    id: "pseudocode",
    name: "Pseudocode",
    category: "prose",
    extensions: [".pseudo", ".algo"],
    mimeType: "text/x-pseudocode"
  }
];

export class PolyglotDetector {
  /**
   * Fast detection of file type & language from path and optional raw content.
   */
  static detect(filepath: string, content?: string): LanguageDefinition {
    const filename = filepath.replace(/\\/g, "/").split("/").pop() ?? "";
    const lowerFilename = filename.toLowerCase();

    // Stage 1: Exact filename match
    for (const lang of LANGUAGE_REGISTRY) {
      if (lang.filenames && lang.filenames.some(fn => fn.toLowerCase() === lowerFilename)) {
        return lang;
      }
    }

    // Stage 2: Shebang check if content is provided
    if (content && content.startsWith("#!")) {
      const firstLine = content.slice(0, content.indexOf("\n") === -1 ? content.length : content.indexOf("\n")).toLowerCase();
      for (const lang of LANGUAGE_REGISTRY) {
        if (lang.shebangs && lang.shebangs.some(sh => firstLine.includes(sh))) {
          return lang;
        }
      }
    }

    // Stage 3: Extension match
    const extMatch = filename.match(/\.[a-zA-Z0-9_-]+$/);
    const ext = extMatch ? extMatch[0].toLowerCase() : "";
    if (ext) {
      for (const lang of LANGUAGE_REGISTRY) {
        if (lang.extensions.includes(ext)) {
          return lang;
        }
      }
    }

    // Stage 4: Content heuristics
    if (content) {
      const trimmed = content.trim();
      if (trimmed.startsWith("<!DOCTYPE html") || trimmed.startsWith("<html")) {
        return LANGUAGE_REGISTRY.find(l => l.id === "html")!;
      }
      if ((trimmed.startsWith("{") && trimmed.endsWith("}")) || (trimmed.startsWith("[") && trimmed.endsWith("]"))) {
        try {
          JSON.parse(trimmed);
          return LANGUAGE_REGISTRY.find(l => l.id === "json")!;
        } catch {
          // not valid JSON
        }
      }
    }

    // Fallback: generic text/plain
    return {
      id: "unknown",
      name: "Plain Text / Generic",
      category: "prose",
      extensions: [ext],
      mimeType: "text/plain"
    };
  }

  /**
   * Structural parser extracting symbols, AST declarations, imports, and metrics.
   */
  static parseCode(filepath: string, content: string): ParseResult {
    const lang = this.detect(filepath, content);
    const lines = content.split(/\r?\n/);
    
    let blankLines = 0;
    let commentLines = 0;
    let codeLines = 0;
    let inBlockComment = false;

    const imports: string[] = [];
    const classes: string[] = [];
    const functions: string[] = [];
    const interfaces: string[] = [];
    const constants: string[] = [];

    // Language-specific patterns
    for (let i = 0; i < lines.length; i++) {
      const rawLine = lines[i]!;
      const line = rawLine.trim();

      if (!line) {
        blankLines++;
        continue;
      }

      // Multi-line comment tracking
      if (inBlockComment) {
        commentLines++;
        if (line.includes("*/") || (lang.id === "python" && line.includes('"""'))) {
          inBlockComment = false;
        }
        continue;
      }

      if (line.startsWith("/*") || (lang.id === "python" && line.startsWith('"""') && !line.slice(3).includes('"""'))) {
        inBlockComment = true;
        commentLines++;
        continue;
      }

      // Single-line comment checks
      if (
        line.startsWith("//") ||
        line.startsWith("#") ||
        line.startsWith(";") ||
        line.startsWith("--")
      ) {
        commentLines++;
        continue;
      }

      codeLines++;

      // Symbol extraction based on language grammar
      switch (lang.id) {
        case "rust": {
          const modMatch = line.match(/^(?:pub\s+)?use\s+([\w:]+)/);
          if (modMatch?.[1]) imports.push(modMatch[1]);
          const fnMatch = line.match(/^(?:pub(?:\([^)]+\))?\s+)?(?:async\s+)?fn\s+([a-zA-Z0-9_]+)/);
          if (fnMatch?.[1]) functions.push(fnMatch[1]);
          const structMatch = line.match(/^(?:pub\s+)?struct\s+([a-zA-Z0-9_]+)/);
          if (structMatch?.[1]) classes.push(structMatch[1]);
          const traitMatch = line.match(/^(?:pub\s+)?trait\s+([a-zA-Z0-9_]+)/);
          if (traitMatch?.[1]) interfaces.push(traitMatch[1]);
          break;
        }
        case "typescript":
        case "javascript": {
          const impMatch = line.match(/^import\s+(?:.*?from\s+)?["'](.*?)["']/);
          if (impMatch?.[1]) imports.push(impMatch[1]);
          const fnMatch = line.match(/^(?:export\s+)?(?:default\s+)?(?:async\s+)?function\s+([a-zA-Z0-9_]+)/);
          if (fnMatch?.[1]) functions.push(fnMatch[1]);
          const clsMatch = line.match(/^(?:export\s+)?(?:abstract\s+)?class\s+([a-zA-Z0-9_]+)/);
          if (clsMatch?.[1]) classes.push(clsMatch[1]);
          const ifMatch = line.match(/^(?:export\s+)?interface\s+([a-zA-Z0-9_]+)/);
          if (ifMatch?.[1]) interfaces.push(ifMatch[1]);
          const constMatch = line.match(/^(?:export\s+)?const\s+([a-zA-Z0-9_]+)/);
          if (constMatch?.[1]) constants.push(constMatch[1]);
          break;
        }
        case "python": {
          const impMatch = line.match(/^(?:from\s+([a-zA-Z0-9_.]+)\s+import|import\s+([a-zA-Z0-9_.]+))/);
          if (impMatch) imports.push(impMatch[1] || impMatch[2] || "");
          const defMatch = line.match(/^(?:async\s+)?def\s+([a-zA-Z0-9_]+)/);
          if (defMatch?.[1]) functions.push(defMatch[1]);
          const clsMatch = line.match(/^class\s+([a-zA-Z0-9_]+)/);
          if (clsMatch?.[1]) classes.push(clsMatch[1]);
          break;
        }
        case "csharp": {
          const usingMatch = line.match(/^using\s+([a-zA-Z0-9_.]+);/);
          if (usingMatch?.[1]) imports.push(usingMatch[1]);
          const clsMatch = line.match(/(?:public|private|internal|protected)?\s*(?:static|sealed|abstract)?\s*class\s+([a-zA-Z0-9_]+)/);
          if (clsMatch?.[1]) classes.push(clsMatch[1]);
          const ifMatch = line.match(/(?:public|private|internal|protected)?\s*interface\s+([a-zA-Z0-9_]+)/);
          if (ifMatch?.[1]) interfaces.push(ifMatch[1]);
          break;
        }
        case "cpp":
        case "c": {
          const incMatch = line.match(/^#include\s*[<"]([^>"]+)[>"]/);
          if (incMatch?.[1]) imports.push(incMatch[1]);
          const clsMatch = line.match(/^class\s+([a-zA-Z0-9_]+)/);
          if (clsMatch?.[1]) classes.push(clsMatch[1]);
          const fnMatch = line.match(/^[a-zA-Z0-9_*&]+\s+([a-zA-Z0-9_]+)\s*\([^)]*\)\s*\{/);
          if (fnMatch?.[1] && !["if", "for", "while", "switch"].includes(fnMatch[1])) {
            functions.push(fnMatch[1]);
          }
          break;
        }
        case "go": {
          const impMatch = line.match(/^import\s+["'](.*?)["']/);
          if (impMatch?.[1]) imports.push(impMatch[1]);
          const fnMatch = line.match(/^func\s+(?:\([^)]+\)\s+)?([a-zA-Z0-9_]+)/);
          if (fnMatch?.[1]) functions.push(fnMatch[1]);
          const typeMatch = line.match(/^type\s+([a-zA-Z0-9_]+)\s+struct/);
          if (typeMatch?.[1]) classes.push(typeMatch[1]);
          const ifMatch = line.match(/^type\s+([a-zA-Z0-9_]+)\s+interface/);
          if (ifMatch?.[1]) interfaces.push(ifMatch[1]);
          break;
        }
        case "assembly": {
          const labelMatch = line.match(/^([a-zA-Z0-9_]+):/);
          if (labelMatch?.[1]) functions.push(labelMatch[1]);
          break;
        }
      }
    }

    return {
      language: lang.name,
      category: lang.category,
      mimeType: lang.mimeType,
      isBinary: false,
      totalLines: lines.length,
      codeLines,
      commentLines,
      blankLines,
      imports: Array.from(new Set(imports)),
      declarations: {
        classes: Array.from(new Set(classes)),
        functions: Array.from(new Set(functions)),
        interfaces: Array.from(new Set(interfaces)),
        constants: Array.from(new Set(constants))
      },
      validationGate: lang.testRunner ? `${lang.buildGate ? lang.buildGate + " && " : ""}${lang.testRunner}` : lang.buildGate
    };
  }
}

File: packages/yolo/scripts/polyglot_detector.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""
Universal Polyglot Language & File Type Detector & Parser (Python 3.10+)
Directly inspired by GitHub Linguist, Google Magika heuristics, and Tree-sitter AST conventions.
"""

import os
import re
import json
from pathlib import Path

REGISTRY = {
    "rust": {
        "name": "Rust",
        "category": "programming",
        "extensions": [".rs"],
        "filenames": ["Cargo.toml"],
        "test_gate": "cargo test",
        "build_gate": "cargo check --workspace --all-targets"
    },
    "typescript": {
        "name": "TypeScript",
        "category": "programming",
        "extensions": [".ts", ".tsx", ".mts", ".cts"],
        "filenames": ["tsconfig.json"],
        "shebangs": ["bun", "deno", "ts-node"],
        "test_gate": "bun test",
        "build_gate": "tsc --noEmit"
    },
    "javascript": {
        "name": "JavaScript",
        "category": "programming",
        "extensions": [".js", ".jsx", ".mjs", ".cjs"],
        "filenames": ["package.json"],
        "shebangs": ["node", "bun"],
        "test_gate": "bun test || npm test",
        "build_gate": "node --check"
    },
    "python": {
        "name": "Python",
        "category": "programming",
        "extensions": [".py", ".pyw", ".pyi"],
        "filenames": ["pyproject.toml", "requirements.txt"],
        "shebangs": ["python", "python3", "uv run"],
        "test_gate": "pytest || python -m unittest",
        "build_gate": "python -m py_compile"
    },
    "csharp": {
        "name": "C#",
        "category": "programming",
        "extensions": [".cs", ".csx"],
        "test_gate": "dotnet test",
        "build_gate": "dotnet build"
    },
    "cpp": {
        "name": "C++",
        "category": "programming",
        "extensions": [".cpp", ".cxx", ".cc", ".hpp", ".hxx"],
        "filenames": ["CMakeLists.txt"],
        "test_gate": "ctest || make test",
        "build_gate": "cmake --build ."
    },
    "c": {
        "name": "C",
        "category": "programming",
        "extensions": [".c", ".h"],
        "filenames": ["Makefile"],
        "test_gate": "make test",
        "build_gate": "make"
    },
    "assembly": {
        "name": "Assembly",
        "category": "programming",
        "extensions": [".asm", ".s", ".nasm"],
        "build_gate": "nasm -f elf64 || as"
    },
    "go": {
        "name": "Go",
        "category": "programming",
        "extensions": [".go"],
        "filenames": ["go.mod"],
        "test_gate": "go test ./...",
        "build_gate": "go build ./..."
    },
    "json": {
        "name": "JSON",
        "category": "data",
        "extensions": [".json", ".jsonc", ".jsonl"]
    },
    "markdown": {
        "name": "Markdown",
        "category": "prose",
        "extensions": [".md", ".markdown"]
    }
}

class PolyglotDetector:
    @staticmethod
    def detect(filepath: str, content: str = "") -> dict:
        p = Path(filepath)
        filename = p.name.lower()
        ext = p.suffix.lower()

        for lang_id, data in REGISTRY.items():
            if "filenames" in data and any(fn.lower() == filename for fn in data["filenames"]):
                return {"id": lang_id, **data}

        if content and content.startswith("#!"):
            first_line = content.splitlines()[0].lower()
            for lang_id, data in REGISTRY.items():
                if "shebangs" in data and any(sh in first_line for sh in data["shebangs"]):
                    return {"id": lang_id, **data}

        for lang_id, data in REGISTRY.items():
            if ext in data.get("extensions", []):
                return {"id": lang_id, **data}

        return {
            "id": "unknown",
            "name": "Plain Text / Generic",
            "category": "prose",
            "extensions": [ext]
        }

    @staticmethod
    def parse_symbols(filepath: str, content: str) -> dict:
        info = PolyglotDetector.detect(filepath, content)
        lang_id = info["id"]
        classes = []
        functions = []
        imports = []

        lines = content.splitlines()
        for line in lines:
            s = line.strip()
            if lang_id == "python":
                if s.startswith("def "):
                    m = re.match(r"^def\s+([a-zA-Z0-9_]+)", s)
                    if m: functions.append(m.group(1))
                elif s.startswith("class "):
                    m = re.match(r"^class\s+([a-zA-Z0-9_]+)", s)
                    if m: classes.append(m.group(1))
                elif s.startswith("import ") or s.startswith("from "):
                    imports.append(s)
            elif lang_id == "rust":
                if "fn " in s:
                    m = re.search(r"\bfn\s+([a-zA-Z0-9_]+)", s)
                    if m: functions.append(m.group(1))
                elif "struct " in s:
                    m = re.search(r"\bstruct\s+([a-zA-Z0-9_]+)", s)
                    if m: classes.append(m.group(1))
                elif s.startswith("use "):
                    imports.append(s)
            elif lang_id in ("typescript", "javascript"):
                if "function " in s:
                    m = re.search(r"\bfunction\s+([a-zA-Z0-9_]+)", s)
                    if m: functions.append(m.group(1))
                elif "class " in s:
                    m = re.search(r"\bclass\s+([a-zA-Z0-9_]+)", s)
                    if m: classes.append(m.group(1))
                elif s.startswith("import "):
                    imports.append(s)

        return {
            "language": info["name"],
            "category": info["category"],
            "classes": classes,
            "functions": functions,
            "imports": imports,
            "total_lines": len(lines)
        }

if __name__ == "__main__":
    import sys
    target = sys.argv[1] if len(sys.argv) > 1 else "scripts/polyglot-detector.ts"
    cnt = Path(target).read_text(encoding="utf-8", errors="ignore") if Path(target).exists() else ""
    res = PolyglotDetector.parse_symbols(target, cnt)
    print(json.dumps(res, indent=2))